### PR TITLE
codec: let it panic when we try to encode a unspecified decimal.

### PR DESCRIPTION
**Pull request by hanfei1991**
If a decimal hold an invalid length, we raise a panic when encoding.

### CODE REVIEW

File: util/codec/decimal.go [hunk 1]
```diff
@@ -31,6 +31,7 @@ func EncodeDecimal(b []byte, d types.Datum) []byte {
 	bin, err := dec.ToBin(precision, frac)
 	if err != nil {
 		log.Errorf("should not happen, precision %d, frac %d %v", precision, frac, err)
+		panic("expected error happened during codec")
```
Comment (XuHuaiyu) [inline util/codec/decimal.go]: may be a more detailed message better?

Comment (shenli) [inline util/codec/decimal.go]: When will this happen?

**Comment by shenli** [comment]
@hanfei1991 CI failed.

File: util/codec/decimal.go (hunk 1 above)
Comment (hanfei1991) [inline util/codec/decimal.go]: This will happen in case of #3694 . This problem will be fixed this week.

Comment (hanfei1991) [inline util/codec/decimal.go]: And we should let it panic when we meet an unexpected bug and don't want to return an error.

File: util/codec/decimal.go [hunk 2]
```diff
@@ -31,7 +31,7 @@ func EncodeDecimal(b []byte, d types.Datum) []byte {
 	bin, err := dec.ToBin(precision, frac)
 	if err != nil {
 		log.Errorf("should not happen, precision %d, frac %d %v", precision, frac, err)
-		return b
+		panic("expected error happened during codec")
```
Comment (winoros) [inline util/codec/decimal.go]: combine the panic message with the log error message?

Comment (hanfei1991) [inline util/codec/decimal.go]: Good suggestion.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by shenli** [comment]
Why should we do this? Have you met any thing wrong associate wit this?

**Comment by hanfei1991** [comment]
@shenli #3694

**Comment by shenli** [comment]
So after this PR. The join statement in #3694 will panic?

**Comment by hanfei1991** [comment]
@shenli yes, you are right

**Comment by shenli** [comment]
So it is better to merge this PR after finishing refactor `Equal` function?

**Comment by hanfei1991** [comment]
@shenli You can make sure when a similar bug is triggered again. Panic is better than bug.

**Comment by shenli** [comment]
OK. We should move quick to fix related bug after merging this PR.

**Pull request merged**